Translate introducing JSX (#15)
reactjs/bg.reactjs.org: content/docs/introducing-jsx.md

---
id: introducing-jsx
title: Представяне на JSX
permalink: docs/introducing-jsx.html
prev: hello-world.html
next: rendering-elements.html
---

Разгледайте тази декларация на променлива:

```js
const element = <h1>Здравей, свят!</h1>;
```

Този забавен синтаксис не е нито низ, нито HTML.

Той се нарича JSX и е синтактично разширение на JavaScript. Препоръчваме да го използвате с React, когато описвате как трябва да изглежда потребителският интерфейс. JSX може да ви напомня за шаблонен език, но идва с всички възможности на JavaScript.

JSX произвежда React "елементи". Ще разгледаме рендирането им в DOM в [следващия раздел](/docs/rendering-elements.html). По-долу можете да намерите основите на JSX, необходими за да започнете.

### Защо JSX? {#why-jsx}

React приема факта, че логиката за рендирането е без съмнение в съчетание с друга логика на потребителския интерфейс: как се обработват събитията, как се променя състоянието във времето и как се подготвят данните за показване.

> Вместо изкуствено разделяне на *технологии* чрез поставяне на markup и логика в отделни файлове, React [разделя *целите*](https://en.wikipedia.org/wiki/Separation_of_concerns) с свободно свързани единици, наречени "компоненти", които съдържат и двете. Ще се върнем към компонентите в [следваща секция](/docs/components-and-props.html), но ако все още не ви е удобно да поставяте markup в JS, [тази презентация](https://www.youtube.com/watch?v=x7cQ3mrcKaY) може да ви убеди в това.

Reviewer: ```suggestion
Вместо изкуствено разделяне на *технологии* чрез поставяне на markup и логика в отделни файлове, React [разделя *отговорностите*](https://en.wikipedia.org/wiki/Separation_of_concerns) чрез свободно свързани единици, наречени "компоненти", които съдържат и двете. Ще се върнем към компонентите в [следващата секция](/docs/components-and-props.html), но ако все още не ви е удобно да поставяте markup в JS, [тази презентация](https://www.youtube.com/watch?v=x7cQ3mrcKaY) може да ви убеди в това.
```

Виждам съм на няколко места "separation of concerns" да се превежда като "разделяне на отговорности" и съм склонен да се съглася. Какво мислиш?

Author: Супер предложение! Благодаря.

> React [не изисква](/docs/react-without-jsx.html) използването на JSX, но повечето хора го намират за полезен като визуална помощ при работа с потребителски интерфейс в JavaScript кода. Той също така позволява на React да показва по-полезни предупреждения и съобщения за грешки.

Reviewer: ```suggestion
React [не изисква](/docs/react-without-jsx.html) използването на JSX, но повечето хора го намират за полезен като визуална помощ при работа с потребителски интерфейс в JavaScript код. Той също така позволява на React да показва по-полезни предупреждения и съобщения за грешки.
```

След като уточнихме това, нека да започнем!

### Вграждане на изрази в JSX {#embedding-expressions-in-jsx}

В примера по-долу, ние декларираме променлива, наречена `name`, и след това я използваме в JSX, като я обгърнем във фигурни скоби:

```js{1,2}
const name = 'Иван Атанасов';
const element = <h1>Здравей, {name}</h1>;

ReactDOM.render(
  element,
  document.getElementById('root')
);
```

> Можете да поставите всички валидни [JavaScript изрази](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#Expressions) в къдравите скоби в JSX. Например, `2 + 2`, `user.firstName` или `formatName (user)` са валидни JavaScript изрази.

Reviewer: ```suggestion
Можете да поставите всeки валиден [JavaScript израз](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#Expressions) в къдравите скоби в JSX. Например, `2 + 2`, `user.firstName` или `formatName(user)` са валидни JavaScript изрази.
```

[](codepen://introducing-jsx)

> Разделяме JSX върху няколко реда за четливост. Въпреки че не е необходимо, когато правите това, препоръчваме да го опаковате в скоби, за да избегнете капаните на [автоматичното въвеждане на точка и запетая](http://stackoverflow.com/q/2846283).

Reviewer: ```suggestion
Разделяме JSX на няколко реда за четливост. Въпреки че не е необходимо, когато правите това, препоръчваме да го заградите в скоби, за да избегнете проблемите с [автоматичното въвеждане на точка и запетая](http://stackoverflow.com/q/2846283).
```

Преди време срещнах една дама, която превеждаше детски книжки от английски на български. По стечение на обстоятелствата имах английска версия на книжка, която тя беше превела и ми беше интересно да ги сравня. С изненада разбрах, че на доста места използва различен изказ и някой от глаголите са заменени с напълно различни думи на български. Резултата беше обаче доста по-добър от колкото да използва английските им еквиваленти. Та от нея знам, че превода трябва да е основно по смисъл а не толкова по синтаксис. "Разделяме JSX _върху_ няколко реда за четливост" звучи малко странно. "pitfalls" действително се превежда като "капани", но използването на тази дума също ми звучи странно тук. 

П.С.
Съвсем спокойно може да игнорираш някои от коментарите ми. Предлагам варианти, които на мен ми звучат по-добре, но не искам да бъда педантичен.

Author: Много си прав и съм напълно съгласен, има голяма разлика между превод и локализация, тоест и да използваш подходящите изрази на определения език. That being said, определено ми е трудно някои неща да ги преведа добре на Български. Може би ще натрупам повече опит с повече преведени статий :)

```

Можете също да използвате фигурни скоби, за да вградите JavaScript израз в атрибут:

```js
const element = <img src={user.avatarUrl}></img>;
```

Не поставяйте кавички около фигурни скоби, когато вграждате JavaScript израз в атрибут. Трябва да използвате кавички (за стойности на низове) или фигурни скоби (за изрази), но не и двете в един и същ атрибут.

>**Внимание:**
>
>Тъй като JSX е по-близо до JavaScript, отколкото до HTML, React DOM използва конвенцията за именуване "camelCase" за имената на атрибутите вместо HTML имената на атрибутите.
>
>Например `class` става [`className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className) в JSX, а `tabindex` става [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/tabIndex).

### Указване на деца с JSX {#specifying-children-with-jsx}

Ако даден таг е празен, можете да го затворите незабавно с `/>`, като в XML:

```js
const element = <img src={user.avatarUrl} />;
```

JSX таговете могат да съдържат деца:

```js
const element = (
  <div>
    <h1>Здравей!</h1>
    <h2>Хубаво е че си тук.</h2>
  </div>
);
```

### JSX предотвратява атаки чрез инжектиране {#jsx-prevents-injection-attacks}

> В JSX е безопасно да се вгради входяща информация от потребител:

Reviewer: ```suggestion
В JSX е безопасно да се използва информация въведена от потребителя:
```

```js
const title = response.potentiallyMaliciousInput;
// Това е безопасно:
const element = <h1>{title}</h1>;
```

> По подразбиране, React DOM [избягва](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html) всички стойности, вградени в JSX, преди да ги рендира. По този начин се гарантира, че никога не можете да инжектирате нещо, което не е изрично написано във вашето приложение. Всичко се преобразува в низ, преди да се визуализира. Това помага да се предотвратят [XSS (cross-site-scripting)](https://en.wikipedia.org/wiki/Cross-site_scripting) атаки.

Reviewer: Тук мисля, че `escapes` трябва да стане `escape-ва` тъй като няма точен превод на български.

### JSX представя обекти {#jsx-represents-objects}

> Babel компилира JSX до `React.createElement()` повиквания.

Reviewer: ```suggestion
Babel компилира JSX до `React.createElement()` извиквания.
```

const element = React.createElement(
  'h1',
  {className: 'greeting'},
  'Здравей, свят!'
);
```

`React.createElement()` извършва няколко проверки, за да ви помогне да пишете безпроблемен код, но в крайна сметка създава обект като този:

```js
// Забележка: тази структура е опростена
const element = {
  type: 'h1',
  props: {
    className: 'greeting',
    children: 'Здравей, свят!'
  }
};
```

> Тези обекти се наричат ​​"React елементи". Можете да ги мислите като за описания на това, което искате да видите на екрана. React чете тези обекти и ги използва, за да построи DOM и да го поддържа актуализиран.

Reviewer: ```suggestion
Тези обекти се наричат ​​"React елементи". Може да мислите за тях като за описания на това, което искате да видите на екрана. React чете тези обекти и ги използва, за да построи DOM и да го поддържа актуализиран.
```

Ще разгледаме визуализирането на React елементи в DOM в следващия раздел.

>**Съвет:**
>

> >Препоръчваме ви да използвате ["Babel" езикова дефиниция](http://babeljs.io/docs/editors) за редактора който сте избрали, така че и ES6 и JSX кодът да е с правилен syntax highlighting. Този уебсайт използва цветовата схема [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/), която е съвместима с тази дефиниция.

Reviewer: ```suggestion
>Препоръчваме ви да използвате ["Babel" езикова дефиниция](http://babeljs.io/docs/editors) за редактора, който сте избрали, така че и ES6 и JSX кодът да е с правилен syntax highlighting. Този уебсайт използва цветовата схема [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/), която е съвместима с тази дефиниция.
```
